### PR TITLE
Make provider cache more resilient to collisions

### LINKED ISSUE #13248: Test failure: Reports_debug_info_for_most_similar_existing_service_provider
This test [fails](http://aspnetci/viewLog.html?buildId=547358&buildTypeId=Releases_22xPublic_OsxUniverse) occasionally with the following error:
```
Assert.Equal() Failure
Expected: 4
Actual:   3
   at Microsoft.EntityFrameworkCore.ServiceProviderCacheTest.Reports_debug_info_for_most_similar_existing_service_provider() in /_/test/EFCore.Tests/ServiceProviderCacheTest.cs:line 190
```
Other tests within that build may have failed with a similar message, but they are not listed here. Check the link above for more info.

This test failed on 2.2.

CC @ajcvickers

This issue was made automatically. If there is a problem contact ryanbrandenburg.

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #13248

### CODE REVIEW

File: src/EFCore/Infrastructure/CoreOptionsExtension.cs [hunk 1]
```diff
@@ -354,9 +354,9 @@ public virtual long GetServiceProviderHashCode()
             {
                 var hashCode = GetLoggerFactory()?.GetHashCode() ?? 0L;
                 hashCode = (hashCode * 397) ^ (GetMemoryCache()?.GetHashCode() ?? 0L);
-                hashCode = (hashCode * 397) ^ _sensitiveDataLoggingEnabled.GetHashCode();
-                hashCode = (hashCode * 397) ^ _richDataErrorHandingEnabled.GetHashCode();
-                hashCode = (hashCode * 397) ^ _warningsConfiguration.GetServiceProviderHashCode();
+                hashCode = (hashCode * 31) ^ _sensitiveDataLoggingEnabled.GetHashCode();
+                hashCode = (hashCode * 17) ^ _richDataErrorHandingEnabled.GetHashCode();
+                hashCode = (hashCode * 3163) ^ _warningsConfiguration.GetServiceProviderHashCode();
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/CoreOptionsExtension.cs]: Why 3163 here?

Comment (ajcvickers) [reply src/EFCore/Infrastructure/CoreOptionsExtension.cs]: Because it's a bigger prime.

Comment (AndriySvyryd) [reply src/EFCore/Infrastructure/CoreOptionsExtension.cs]: If you want to reduce `warningsConfiguration` hash code impact on collisions you can use `3` for the previous two as they are just bools and use a bigger prime for the last one

Comment (AndriySvyryd) [reply src/EFCore/Infrastructure/CoreOptionsExtension.cs]: Like `1073742113`, it's just below 3^31

Comment (ajcvickers) [reply src/EFCore/Infrastructure/CoreOptionsExtension.cs]: Will do.

**Pull request merged**